interpreter/ruby/ruby.go

```diff
@@ -628,57 +724,376 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+		log.Warnf("RubySymbolizer: Failed to get source label (iseq@0x%08x) %d %08x, %v", iseqBody, frameAddrType, frameFlags, err)
+		return &rubyIseq{}, err
 	}
 
-	funcNamePtr := r.rm.Ptr(iseqBody +
-		libpf.Address(vms.iseq_constant_body.location+vms.iseq_location_struct.base_label))
-	functionName, err := r.getStringCached(funcNamePtr, r.readRubyString)
+	iseqBaseLabelPtr := npsr.Ptr(dataBytes, uint(vms.iseq_location_struct.base_label))
+	iseqBaseLabel, err := r.getStringCached(iseqBaseLabelPtr, r.readRubyString)
 	if err != nil {
-		return err
+		log.Warnf("RubySymbolizer: Failed to get source base label (iseq@0x%08x) %d %08x, %v", iseqBody, frameAddrType, frameFlags, err)
+		return &rubyIseq{}, err
 	}
 
-	// Ruby doesn't provide the information about the function offset for the
-	// particular line. So we report 0 for this to our backend.
+	// Body used for for qualified method label is indirect, need to do: iseq body -> local iseq -> iseq body
+	// https://github.com/ruby/ruby/blob/v3_4_5/vm_backtrace.c#L1943
+	// https://github.com/ruby/ruby/blob/v3_4_5/iseq.c#L1426
+	localIseqPtr := r.rm.Ptr(iseqBody + libpf.Address(vms.iseq_constant_body.local_iseq))
+	iseqLocalBody := r.rm.Ptr(localIseqPtr + libpf.Address(vms.iseq_struct.body))
+
+	// Check iseq body type to see if it is a method before trying to read it
+	// https://github.com/ruby/ruby/blob/v3_4_5/iseq.c#L1428-L1430
+	iseqType := r.rm.Uint32(iseqLocalBody + libpf.Address(vms.iseq_constant_body.iseq_type))
+
+	var methodName libpf.String
+	if iseqType == iseqTypeMethod {
+		methodNamePtr := r.rm.Ptr(iseqLocalBody +
+			libpf.Address(vms.iseq_constant_body.location+vms.iseq_location_struct.base_label))
+		methodName, err = r.getStringCached(methodNamePtr, r.readRubyString)
+		if err != nil {
+			log.Warnf("Unable to find local method name on iseq method (%d) (iseq@0x%08x) %v", iseqType, iseqBody, err)
+		}
+	}
+
+	return &rubyIseq{
+		label:          iseqLabel,
+		baseLabel:      iseqBaseLabel,
+		methodName:     methodName,
+		sourceFileName: sourceFileName,
+		line:           libpf.SourceLineno(lineNo),
+	}, nil
+}
+
+func (r *rubyInstance) Symbolize(frame *host.Frame, frames *libpf.Frames) error {
+	if !frame.Type.IsInterpType(libpf.Ruby) {
+		return interpreter.ErrMismatchInterpreterType
+	}
+	sfCounter := successfailurecounter.New(&r.successCount, &r.failCount)
+	defer sfCounter.DefaultToFailure()
+
+	var err error
+	var iseqBody libpf.Address
+	var classPath libpf.String
+	var methodName libpf.String
+	var fullLabel libpf.String
+	var sourceFile libpf.String
+	var sourceLine libpf.SourceLineno
+	var singleton bool
+	var cframe bool
+	var cme bool
+
+	vms := &r.r.vmStructs
+	frameAddr := libpf.Address(frame.File & support.RubyAddrMask48Bit)
+	frameAddrType := uint8(frame.File >> 48)
+	pc := libpf.Address(frame.Lineno & support.RubyAddrMask48Bit)
+
+	frameFlags := unpackEnvFlags(uint16(frame.Lineno >> 48))
+
+	switch frameAddrType {
+	case support.RubyFrameTypeCmeCfunc:
```

Comment:
since you asked for nits - I wonder if it wold read better if we create separate functions for each case. I think each case is independent.